Integrate NNC conv2d with fuser

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55213 Integrate NNC conv2d with fuser**
* #55025 [nnc] Move device type checks from isSupported to typesAreSupported
* #54920 NNC depthwise conv2d implementation

Adds the integration of conv2d with the TE fuser.  A few things of interest:

- I'm *super* selective of what convs get lowered.  Only 3x3 depthwise, because
  I've benchmarked those to death and I'm pretty sure it's a good change.

- I'm allowing single-node "fusion" groups for supported convs.  (Maybe this is
  a sign that conv2d codegen should go through a different path entirely, but
  it seems to basically work).

I'll shared full benchmarkr results once I clean them up a little.  To
summarize, I tested the following torchvision models containing depthwise
convolutions.  Results are single-core on a skylake-avx512:

mobilenet_v2: 8% improvement
mobilenet_v3: 9% improvement
mnasnet: 10% improvement
shufflenet: 18% improvement

Note these are comparing against a baseline with a fast-but-buggy grouped
convolution implementation in MKLDNN.  So perf results will be better if
compared on master, but I'm going to assume the MKLDNN bug will be fixed and
re-enabled.

Perf results are more complicated when comparing to freezing plus conversion to
mkldnn layout; mobilenet v2/v3 are still faster, but mnasnet and shufflenet are
not.  Landing this doesn't prevent MKLDNN freezing from kicking in though, so
there's no harm (although landing mkldnn freezing will regress mobilenet, but
cest la vie).

Differential Revision: [D27530272](https://our.internmc.facebook.com/intern/diff/D27530272/)

bertmaher committed:
Integrate NNC conv2d with fuser

Adds the integration of conv2d with the TE fuser.  A few things of interest:

- I'm *super* selective of what convs get lowered.  Only 3x3 depthwise, because
  I've benchmarked those to death and I'm pretty sure it's a good change.

- I'm allowing single-node "fusion" groups for supported convs.  (Maybe this is
  a sign that conv2d codegen should go through a different path entirely, but
  it seems to basically work).

I'll shared full benchmarkr results once I clean them up a little.  To
summarize, I tested the following torchvision models containing depthwise
convolutions.  Results are single-core on a skylake-avx512:

mobilenet_v2: 8% improvement
mobilenet_v3: 9% improvement
mnasnet: 10% improvement
shufflenet: 18% improvement

Note these are comparing against a baseline with a fast-but-buggy grouped
convolution implementation in MKLDNN.  So perf results will be better if
compared on master, but I'm going to assume the MKLDNN bug will be fixed and
re-enabled.

Perf results are more complicated when comparing to freezing plus conversion to
mkldnn layout; mobilenet v2/v3 are still faster, but mnasnet and shufflenet are
not.  Landing this doesn't prevent MKLDNN freezing from kicking in though, so
there's no harm (although landing mkldnn freezing will regress mobilenet, but
cest la vie).

Differential Revision: [D27530272](https://our.internmc.facebook.com/intern/diff/D27530272/)

[ghstack-poisoned]

bertmaher committed:
TORCH_API on "Integrate NNC conv2d with fuser"


Adds the integration of conv2d with the TE fuser.  A few things of interest:

- I'm *super* selective of what convs get lowered.  Only 3x3 depthwise, because
  I've benchmarked those to death and I'm pretty sure it's a good change.

- I'm allowing single-node "fusion" groups for supported convs.  (Maybe this is
  a sign that conv2d codegen should go through a different path entirely, but
  it seems to basically work).

I'll shared full benchmarkr results once I clean them up a little.  To
summarize, I tested the following torchvision models containing depthwise
convolutions.  Results are single-core on a skylake-avx512:

mobilenet_v2: 8% improvement
mobilenet_v3: 9% improvement
mnasnet: 10% improvement
shufflenet: 18% improvement

Note these are comparing against a baseline with a fast-but-buggy grouped
convolution implementation in MKLDNN.  So perf results will be better if
compared on master, but I'm going to assume the MKLDNN bug will be fixed and
re-enabled.

Perf results are more complicated when comparing to freezing plus conversion to
mkldnn layout; mobilenet v2/v3 are still faster, but mnasnet and shufflenet are
not.  Landing this doesn't prevent MKLDNN freezing from kicking in though, so
there's no harm (although landing mkldnn freezing will regress mobilenet, but
cest la vie).

Differential Revision: [D27530272](https://our.internmc.facebook.com/intern/diff/D27530272/)

[ghstack-poisoned]

bertmaher committed:
rebase on "Integrate NNC conv2d with fuser"


Adds the integration of conv2d with the TE fuser.  A few things of interest:

- I'm *super* selective of what convs get lowered.  Only 3x3 depthwise, because
  I've benchmarked those to death and I'm pretty sure it's a good change.

- I'm allowing single-node "fusion" groups for supported convs.  (Maybe this is
  a sign that conv2d codegen should go through a different path entirely, but
  it seems to basically work).

I'll shared full benchmarkr results once I clean them up a little.  To
summarize, I tested the following torchvision models containing depthwise
convolutions.  Results are single-core on a skylake-avx512:

mobilenet_v2: 8% improvement
mobilenet_v3: 9% improvement
mnasnet: 10% improvement
shufflenet: 18% improvement

Note these are comparing against a baseline with a fast-but-buggy grouped
convolution implementation in MKLDNN.  So perf results will be better if
compared on master, but I'm going to assume the MKLDNN bug will be fixed and
re-enabled.

Perf results are more complicated when comparing to freezing plus conversion to
mkldnn layout; mobilenet v2/v3 are still faster, but mnasnet and shufflenet are
not.  Landing this doesn't prevent MKLDNN freezing from kicking in though, so
there's no harm (although landing mkldnn freezing will regress mobilenet, but
cest la vie).

Differential Revision: [D27530272](https://our.internmc.facebook.com/intern/diff/D27530272/)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -627,7 +607,10 @@ class TensorExprFuser {
   // until there is nothing we can pull in.
   std::pair<graph_node_list::iterator, bool> createFusionGroup(
       Node* fusion_node) {
-    if (min_group_size_ == 1) {
```
Unrelated to this PR, but maybe we should enable min_group_size == 1 by default? IIRC, @Chillee experiments showed that the perf was reasonable

eellison replied:
Not sure about whether we should do it now, but definitely something to reconsider when we support dynamic shapes

ZolotukhinM reviewed (approved):
Looks reasonable to me, but I'd like to hear what @eellison and @Krovatkin think about it as well.

bertmaher committed:
rebase on "Integrate NNC conv2d with fuser"


Adds the integration of conv2d with the TE fuser.  A few things of interest:

- I'm *super* selective of what convs get lowered.  Only 3x3 depthwise, because
  I've benchmarked those to death and I'm pretty sure it's a good change.

- I'm allowing single-node "fusion" groups for supported convs.  (Maybe this is
  a sign that conv2d codegen should go through a different path entirely, but
  it seems to basically work).

I'll shared full benchmarkr results once I clean them up a little.  To
summarize, I tested the following torchvision models containing depthwise
convolutions.  Results are single-core on a skylake-avx512:

mobilenet_v2: 8% improvement
mobilenet_v3: 9% improvement
mnasnet: 10% improvement
shufflenet: 18% improvement

Note these are comparing against a baseline with a fast-but-buggy grouped
convolution implementation in MKLDNN.  So perf results will be better if
compared on master, but I'm going to assume the MKLDNN bug will be fixed and
re-enabled.

Perf results are more complicated when comparing to freezing plus conversion to
mkldnn layout; mobilenet v2/v3 are still faster, but mnasnet and shufflenet are
not.  Landing this doesn't prevent MKLDNN freezing from kicking in though, so
there's no harm (although landing mkldnn freezing will regress mobilenet, but
cest la vie).

Differential Revision: [D27530272](https://our.internmc.facebook.com/intern/diff/D27530272/)

[ghstack-poisoned]
